crates/ra_fmt2/src/edit_tree.rs

```diff
@@ -0,0 +1,115 @@
+use crate::dsl::{SpacingDsl, SpacingRule, SpaceLoc, SpaceValue};
+use crate::pattern::{Pattern, PatternSet};
+use crate::rules::spacing;
+use crate::trav_util::{walk, walk_nodes, walk_tokens};
+
+use ra_syntax::{
+    NodeOrToken, SmolStr, SyntaxElement,
+    SyntaxKind::{self, *},
+    SyntaxNode, SyntaxToken, TextRange, TextUnit, WalkEvent, T,
+};
+
+use std::collections::{HashMap, HashSet};
+
+// TODO make more like intellij's fmt model
+// Model holds immutable tree and mutable intermediate model to produce diff
+// the model will probably have to create its own tree to add the extra
+// info to each token/node:
+//
+// [1,2,3];
+// can be Brace token, ident, comma all of which knows their own rules and apply
+// them accordingly to produce [1, 2, 3]; ???
+
+#[derive(Debug)]
+// this will probably be an enum that holds a struct for space/indent/ect
+struct AtomEdit {
+    space_loc: SpaceLoc,
+    val: SpaceValue,
+    start: TextUnit,
+}
+
+impl AtomEdit {}
+
+#[derive(Debug)]
+pub(crate) struct SynBlock {
+    //indent: some enum?
+    kind: SyntaxKind,
+    node: SyntaxNode,
+    tokens: Vec<SyntaxToken>,
```

Comment:
I think rather than holding a list of tokens, we should hold a list of child `SynBlock`s.

Reply:
So something like this
```rust
pub(crate) struct SynBlock {
    kind: SyntaxKind,
    parent: SyntaxNode,
    tokens: Vec<SynBlock>,
    range: TextRange,
    prev_whitespace: Option<SyntaxToken>,
}
```
Each token is turned into a SynBlock or every SyntaxElement is a SynBlock.

Reply:
>every SyntaxElement is a SynBlock.

Yeah, every **non-whitespace** element is a block. 

The struct looks almost right two me, though I'd do two changes:

* instead of storing a *parent*, I'd store *this* node (`node: SyntaxElement`), and remove `kind`
* whitespace should be a dedicated type: `prev_whitespace: Whitespace`, something like

```
struct Whitespace {
    original: SyntaxToken,
    indent_spaces: u32,
    additional_spaces: u32, // not sure if we'll need this one
}
```

Reply:
Cool glad to get the initial structuring out of the way now I can do some real damage :laughing: .  Thank you for all the input and help